### PR TITLE
Paul/kado

### PR DESCRIPTION
**Pull request by paullinator**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

 none

### Requirements

If you have made **any** visual changes to the GUI. Make sure you have:

- [x] Tested on iOS device
- [ ] Tested on Android device
- [x] Tested on small-screen device (iPod Touch)
- [ ] Tested on large-screen device (tablet)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206480245141231

### CODE REVIEW

**Review by swansontec — CHANGES_REQUESTED** [review]

File: src/plugins/gui/providers/kadoProvider.ts [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+          allowedCurrencyCodes.crypto[pluginId] = []
+          const tokens = allowedCurrencyCodes.crypto[pluginId]
+          tokens.push({ tokenId: null, otherInfo: { symbol: PLUGINDS_TO_SYMBOL[pluginId] } })
+
+          if (PLUGINIDS_WITH_TOKENS[pluginId]) {
+            for (const asset of blockchain.associatedAssets) {
+              if (!asset.liveOnRamp) continue
+              if (asset.address != null && asset.address !== '0x0000000000000000000000000000000000000000') {
+                if (asset.rampProducts != null && asset.rampProducts.includes(direction)) {
+                  const tokenId = asset.address.toLowerCase().replace('0x', '')
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/providers/kadoProvider.ts:491]: Already done in a new commit that addresses breaking API changes from Kado

File: src/plugins/gui/providers/kadoProvider.ts [hunk 2]
```diff
@@ -425,27 +591,206 @@ export const kadoProvider: FiatProviderFactory = {
...
+                      if (!response.ok) {
+                        const text = await response.text()
+                        console.warn(`Error fetching kado blockchains: ${text}`)
+                        return allowedCurrencyCodes
+                      }
+                      const result = await response.json()
+
+                      const orderInfo = asOrderInfo(result)
+                      if (!orderInfo.success) {
+                        await showUi.showError(new Error(lstrings.fiat_plugin_sell_failed_try_again + '\n\norderInfo.success=false'))
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/plugins/gui/providers/kadoProvider.ts:645]: Unfortunately, the `showUi.showError` API isn't exactly the same as showError. That would require a bit of a refactor which I would agree with but would rather do as another commit/PR.

File: src/components/ui4/FiatExchangeDetailsCard.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  let sourceCurrencyCode = ''
+  let destinationCurrencyCode = ''
+  if (direction === 'buy') {
+    sourceCurrencyCode = fiatAsset.fiatCurrencyCode
+    destinationCurrencyCode = currencyCode
+  } else {
+    sourceCurrencyCode = currencyCode
+    destinationCurrencyCode = fiatAsset.fiatCurrencyCode
+  }
+  useAsyncEffect(
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/components/ui4/FiatExchangeDetailsCard.tsx:115]: I wasn't aware of `useAsyncValue` but good to know. I do like that `useAsyncEffect` allows a tag to help find crashes.

**Comment by paullinator** [comment]
Please check the fixups AND the last two commits which were added since the first review

**Review by swansontec — APPROVED** [review]

**Pull request merged**